packages/ra-core/src/reducer/admin/references/oneToMany.ts

```diff
@@ -114,7 +138,41 @@ export const getReferencesByIds = (
     return Object.keys(references).length > 0 ? references : null;
 };
 
-export const nameRelatedTo = (reference, id, resource, target, filter = {}) => {
+const getRelatedReferences = (previousState: State, resource: string) =>
+    Object.keys(previousState).filter(key => key.includes(resource));
+
+const removeDeletedReferences = (removedIds: Identifier[]) => (
+    previousState: State,
+    key: string
+) => {
+    const idsToRemove = previousState[key].ids.filter(id =>
```

Comment:
instead of filtering twice just to count the number of elements to filter out, why don't you just count once and then compare the length to the length of the original key?

```js
    const idsToKeep = previousState[key].ids.filter(
        id => !removedIds.includes(id)
    );
    if (idsToKeep.length === previousState[key].ids.length) {
        return previousState;
    }
    return {
        ...previousState,
        [key]: {
            ids: idsToKeep,
            total: idsToKeep.length,
        },
    };
```